File: clientapp/lib/domain/service/service.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:servicegeniestore/domain/service/date_time_slots.dart';
import 'package:servicegeniestore/domain/service/service_price.dart';
import 'package:servicegeniestore/domain/service/service_property.dart';
import 'package:servicegeniestore/domain/service/service_state.dart';
import 'package:servicegeniestore/domain/service/time_slot.dart';
import 'package:servicegeniestore/domain/shared/localized_string.dart';

part 'service.g.dart';

@JsonSerializable()
class Service {
  final String id;
  final LocalizedString name;
  final LocalizedString description;
  final LocalizedString? tos;
  final String imageUrl;
  final ServiceState state;
  final List<ServicePrice> price;
  final String categoryId;
  final List<ServiceProperty> properties;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<String> supportedPaymentMethods;
  final List<TimeSlot> defaultTimeSlots;
  final List<DateTimeSlots> specialDates;

  Service({
    required this.id,
    required this.name,
    required this.description,
    this.tos,
    required this.imageUrl,
    required this.state,
    required this.price,
    required this.categoryId,
    required this.properties,
    required this.createdAt,
    required this.updatedAt,
    required this.supportedPaymentMethods,
    required this.defaultTimeSlots,
    required this.specialDates,
  });

  factory Service.fromJson(Map<String, dynamic> json) =>
      _$ServiceFromJson(json);

  Map<String, dynamic> toJson() => _$ServiceToJson(this);
}

File: admin-dashboard-remix/app/domain/services/Service.ts
import { z } from "zod";
import LocalizedStringSchema from "@/domain/models/LocalizedString";
import ServicePriceSchema from "@/domain/models/ServicePrice";

const ServiceStateSchema = z.enum(["public", "private"]);
const ServicePropertySchema = z.union([
  z.object({
    id: z.string().uuid(),
    type: z.literal("text"),
    name: LocalizedStringSchema,
    options: z.array(LocalizedStringSchema),
  }),
  z.object({
    id: z.string().uuid(),
    type: z.literal("number"),
    name: LocalizedStringSchema,
    options: z.array(z.number()),
  }),
]);


const ServiceSchema = z.object({
  id: z.string().uuid(),
  name: LocalizedStringSchema,
  description: LocalizedStringSchema,
  tos: LocalizedStringSchema.optional(),
  image: z.string(),
  state: ServiceStateSchema,
  price: z.array(ServicePriceSchema),
  categoryId: z.string().uuid(),
  properties: z.array(ServicePropertySchema),
  createdAt: z.date(),
  updatedAt: z.date(),
  supportedPaymentMethods: z.array(z.string().uuid()),
});

type Service = z.infer<typeof ServiceSchema>

export default ServiceSchema;

export {
  ServicePropertySchema,
  ServiceStateSchema,
};

export type {
  Service,
};